### ISSUE TITLE
Custom dashboards

### ISSUE THREAD

**Issue by brian428** [ozone-development/ozp-webtop#525]
Looking at the code, it appears that the two current dashboard types (grid and desktop) are essentially "hardcoded" into the webtop. Are there any plans to allow the use of additional custom dashboards without needing to modify the webtop source itself? I'm thinking of things from OWF like border layout, tabs, and fit, as well as customized behavior that goes beyond the layouts that were built into OWF.

**Comment by brian428** [comment]
Just to add, digging deeper into the webtop code, it looks like it would take fairly extensive modifications for us to add a custom dashboard layout. Obviously, forking the code and making significant modifications would make merging future updates a real problem.

While it would take some rework of the current webtop code to wrap up and encapsulate the layout-related logic to support easy addition of custom layouts, I think this is a pretty crucial feature to support. We have several OWF projects that we're planning to move to OZP, but the restriction of only having grid and desktop layouts is proving to be a major obstacle. Thoughts?

**Comment by tonyalletag** [comment]
@brian428 It is great to hear a voice from out side the project. 

The project has plans for  a feature called AppBuilder.  The full concept of this feature is still being worked out but customization of the dashboards has been discussed by the team. @TheAndyStumpf  may have more details on AppBuilder
With the OZP Platform user's are encourage to develop their own custom Webtops if they desire. @arw180 or @emosher may be able to provide more details on custom Webtops

Agreed that forking the code and making significant modification is not ideal as this would most likely create incompatible versions with future releases. We look forward to reviewing pull request from the community to be included in the baseline.

Please continue ping us with good questions and keep this conversation going as we can use it to bump some of these feature request up in our backlog.

**Comment by brian428** [comment]
I'd be interested to hear more about the AppBuilder. That may help with some scenarios. But I'm not sure that a general tool will really cover the full range of needs, since such a tool could never anticipate every dashboard feature that someone may need.

I'm really just picturing some array of dashboard template URLs that can be configured. The webtop would then use those to populate the list of dashboard choices. The template would specify the controller, so from that point a developer can do pretty much anything they want to. All we'd need to do is make sure any custom module scripts are loaded. Ideally, these would also be configurable (to avoid needing to modify the webtop index.html to include additional scripts). But even if that piece required modifying the index.html, that's a much simpler change to manage in terms of merging future webtop updates.

The overall problem, as far as I can tell, is that dashboard-related logic is really spread through the webtop codebase (particularly app.js, the toolbar, chrome, and models.js). There are even places where the possible dashboard values are explicitly limited to only "grid" and "desktop". So I'm not sure how feasible any customization will be until most of that is encapsulated out into separate files, along with some sort of API that supports plugging in additional options? Basically, grid and desktop layout would be build just like any other custom dashboards, only they'd happen to be included with the webtop by default.

Ideally, there would be base classes that can be extended or interfaces that can be implemented to ease the process of creating additional dashboards. But it doesn't look like the code currently uses a class system, and I'm not sure if you have plans (or the inclination) to add one.

Just to throw this out there as well: I've been doing a fair bit of development using TypeScript in my Angular apps. There could be some benefits to switching the project over to TypeScript: first, base classes and interfaces become available, and second, since Angular 2 will be strongly encouraging TypeScript, it would help ease any future Angular 2 conversion work. I've personally been moving everything that I can out of Angular controllers, scopes, and services, and into TS classes. That way, when the Angular 2 changes come through, most of the work will isolated to "plumbing" (how data gets into and out of the views, how events flow, and so forth).

Anyway, I realize this is a lot to digest, so I'll stop for now. Let me know what you think about any or all of this.

**Comment by emosher** [comment]
Before I say anything - **Disclaimer: I am not actively working on the project anymore, so my information may be out of date, or incorrect.  I also don't want to claim to speak for the project or anyone actively working it.**

Now that's out of the way, I think @brian428 brings up a really interesting point.  Namely the ability to configure webtop to behave the way you want.  This kind of hits at what I was envisioning when I talked about a "webtop SDK" with @TheAndyStumpf, @tonyalletag, and @arw180.  The ability for me, as a developer wanting to use OZP, being able to quickly and easily spin up what would be my 'version' of the webtop would be huge.  The amount of people using (read: deploying) OZP is huge and the team cannot possibly anticipate every case that will come up.  Something like this would make it possible for people to leverage the work that has been done here while still creating their "own" webtop.  I think that the idea of "many webtops" is a great one, but developers not currently building OZP tend to view OZP more as a barrier rather than an advantage.  This is mostly because they do not understand it, and I think this could be one way to entice developers to use OZP.  The many webtops idea is more possible than ever with [the new IWC](https://github.com/ozone-development/ozp-iwc), and potential developers shouldn't feel the need to build everything from scratch (or by fork - essentially the same thing).

**Re Appbuilder** - the idea was to have webtop be the "simple" layout management, and appbuilder be the advanced layout manager.  Simple, as I understood it, meant grid and desktop because that was the direction OZP was going.  Advanced was reserved for tabs, accordion, etc.  With the suggestion of @brian428, this could potentially mean that webtop and appbuilder would share most of the same code with only some modules being different.  That would be a great reuse and a great way to lower the barrier to entry for OZP even further.  

All of that said, I think the challenge will be ensuring that the program meets its deadlines and things that have been promised.  There are some great ideas in here, but it represents a possible refactor of the code.  In my experience large refactors are difficult to justify to management.  In a perfect open source utopia, this could be built by the community and adopted by the OZP team.  But, alas this is not a perfect world.  :)

**TL;DR** I like it, I would use hell out of it.  The challenge will be justifying/communicating to management.

**Comment by nialscorva** [comment]
A webtop is just an application that happens to contain other applications inside of iframes.  The simplest possible "Dashboard" would look like:

```
<html>
<body>
  <iframe src="http://widget.example.com"></iframe>
  <iframe src="http://anotherWidget.example.com></iframe>
</body>
</html>
```

I'm not even remotely suggesting that this is a solution for end users, but it's potentially a viable solution if you need a fixed layout with fixed contents for a large audience.

**Comment by brian428** [comment]
Jason, I'm not really sure what you're getting at here. Yes, I know the actual markup for a custom dashboard can be as simple or complex as desired. The issue is all the underlying JS logic and the fact that the webtop app itself is hardcoded to allow only the grid and desktop dashboards. Unless the set of allowable dashboards (and their associated JS logic) becomes extensible, I don't see any way for a custom dashboard to even be an option.

**Comment by emosher** [comment]
I think what @brian428 is getting at here is that he wants use the work that has been done on webtop without having to rewrite it himself.  That is largely what I would want as well.  Everyone wants a 'dashboard', but personally I want to focus on the features needed for my customer not dashboard logic.  If there was a way to easily configure or extend this webtop to do what I want, that would prevent me from having to do it all myself.  I think that's the intent of this thread, but please correct me if I'm wrong...

**Comment by brian428** [comment]
Yes, exactly. I'm dealing with large government projects where the grid and desktop layouts are simply insufficient for what they need. Which means that currently my only two options are to build my own webtop from scratch, or fork the OZP webtop, change it significantly, and deal with ongoing code merge issues forever.

Ideally, the "official" OZP webtop could be reworked to allow layouts to be plugged into it, and to treat the grid and desktop layouts just like any other pluggable layout, except that they happen to be included by default. This would really increase the value of OZP, because it would give developers the freedom to create custom layouts while still leveraging the capabilities of the OZP webtop.

**Comment by arw180** [comment]
The idea of 'pluggable dashboard layouts' in webtop sounded enticing to me at first too, but as of late I've had a change of heart on the issue, and here's why:

The vast majority of the OZP webtop's code deals very much with the two layouts (grid and desktop) themselves. Yes, there's a global toolbar with some links at the top, and there is certainly a bit of logic that could potentially be reused, but by far and large, the application does very little other than manage those two layouts. If we were to support some notion of pluggable dashboard layouts, you would end up taking a lot of baggage with you that you probably don't even want. You'd also (likely) be forced into using our stack (currently AngularJS 1.2.x, Bootstrap 3.2, etc). In addition, we'd have to provide a very large amount of documentation on how to interact with "APIs" that were not designed to be used externally, and thus are probably overly complicated or simply underpowered at the moment.

I think a much more palatable solution for everyone would likely be some documentation and examples showing how to make your own custom "webtops". I believe Jason was trying to make the point that, unlike OWF, a webtop can be extremely lightweight in OZP - there's really no magic to speak of. Of course his six line example won't get you very far, but what if we provided some documentation and examples for creating a rather barebones webtop that plugged into OZP (via IWC) in a very similar way that our 'official' one does? That way, you'd be free to use whatever tech stack you wanted, and you wouldn't have to carry around additional bloat of webtop features you might not need or want, both from a performance and a complexity perspective.

This approach would also allow you to list your webtop in the Apps Mall to share it with others, taking advantage of all the existing infrastructure that is already in place. Sharing pluggable dashboards would require a brand new mechanism. 

I imagine the biggest challenge with this would be the fact that you'd actually have to host the application somewhere and create a Listing for it, as opposed to piggybacking off of the existing OZP Webtop application.
